l1-contracts/test/Rollup.t.sol

```diff
@@ -53,26 +53,67 @@ contract RollupTest is DecoderBase {
     txsHelper = new TxsDecoderHelper();
   }
 
-  function testMixedBlock() public {
-    _testBlock("mixed_block_0");
+  function testMixedBlock(bool _toProve) public {
+    _testBlock("mixed_block_1", _toProve);
+
+    assertEq(rollup.pendingBlockCount(), 2, "Invalid pending block count");
+    assertEq(rollup.provenBlockCount(), _toProve ? 2 : 1, "Invalid proven block count");
   }
 
-  function testConsecutiveMixedBlocks() public {
-    _testBlock("mixed_block_0");
-    _testBlock("mixed_block_1");
+  function testConsecutiveMixedBlocks(uint256 _blocksToProve) public {
+    uint256 toProve = bound(_blocksToProve, 0, 2);
+
+    _testBlock("mixed_block_1", toProve > 0);
+    _testBlock("mixed_block_2", toProve > 1);
+
+    assertEq(rollup.pendingBlockCount(), 3, "Invalid pending block count");
+    assertEq(rollup.provenBlockCount(), 1 + toProve, "Invalid proven block count");
   }
 
-  function testEmptyBlock() public {
-    _testBlock("empty_block_0");
+  function testConsecutiveMixedBlocksNonSequentialProof() public {
+    _testBlock("mixed_block_1", false);
+    _testBlock("mixed_block_2", true);
+
+    assertTrue(rollup.isBlockProven(2), "Block 2 is not proven");
+
+    assertEq(rollup.pendingBlockCount(), 3, "Invalid pending block count");
+    assertEq(rollup.provenBlockCount(), 1, "Invalid proven block count");
   }
 
-  function testConsecutiveEmptyBlocks() public {
-    _testBlock("empty_block_0");
-    _testBlock("empty_block_1");
+  function testEmptyBlock(bool _toProve) public {
+    _testBlock("empty_block_1", _toProve);
+    assertEq(rollup.pendingBlockCount(), 2, "Invalid pending block count");
+    assertEq(rollup.provenBlockCount(), _toProve ? 2 : 1, "Invalid proven block count");
+  }
+
+  function testConsecutiveEmptyBlocks(uint256 _blocksToProve) public {
+    uint256 toProve = bound(_blocksToProve, 0, 2);
+    _testBlock("empty_block_1", toProve > 0);
+    _testBlock("empty_block_2", toProve > 1);
+
+    assertEq(rollup.pendingBlockCount(), 3, "Invalid pending block count");
+    assertEq(rollup.provenBlockCount(), 1 + toProve, "Invalid proven block count");
+  }
+
+  function testRevertInvalidBlockNumber() public {
+    DecoderBase.Data memory data = load("empty_block_1").block;
+    bytes memory header = data.header;
+    bytes32 archive = data.archive;
+    bytes memory body = data.body;
+
+    assembly {
+      // TODO: Hardcoding offsets in the middle of tests is annoying to say the least.
```

Comment:
Could we decode the header, set the block num, then re-encode it?

Reply:
It is not really decoded here, we are providing a json that have both the raw bytes and a decoded version. The thing we are testing is partially that all of the decoding is working as expected and matching the typescript, so I would prefer not to rely on an extra implementation for the correctness.